### PR TITLE
Add catalog

### PR DESCRIPTION
**Pull request by vbauzysvmware**
Added implementation for catalog recourse management.

Reference: #94

### CODE REVIEW

File: vcd/resource_vcd_catalog.go [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+	"log"
+)
+
+func resourceVcdCatalog() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceVcdCatalogCreate,
+		Delete: resourceVcdCatalogDelete,
+		Read:   resourceVcdCatalogRead,
+
+		//resource "vcd_catalog" "OperatingSystems" {
```
Comment (dataclouder) [inline vcd/resource_vcd_catalog.go]: Please remove these comments. They add nothing to the code.

File: vcd/resource_vcd_catalog.go [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+				Optional: true,
+				ForceNew: true,
+			},
+			"name": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"description": &schema.Schema{
```
Comment (dataclouder) [inline vcd/resource_vcd_catalog.go]: Description should be optional

File: vcd/resource_vcd_catalog.go [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+	}
+
+	adminCatalog, err := adminOrg.FindAdminCatalog(d.Id())
+	if err != nil {
+		log.Printf("[DEBUG] Unable to find catalog. Removing from tfstate")
+		d.SetId("")
+		return nil
+	}
+
+	err = adminCatalog.Delete(d.Get("force").(bool), d.Get("recursive").(bool))
```
Comment (dataclouder) [inline vcd/resource_vcd_catalog.go]: What happens if `force` and `recursive` are false? 
How are we dealing with a potential non-deletion of the resource due to these options?

File: vcd/resource_vcd_catalog.go (hunk 1 above)
Comment (vbauzysvmware) [reply vcd/resource_vcd_catalog.go]: done

File: vcd/resource_vcd_catalog.go (hunk 2 above)
Comment (vbauzysvmware) [reply vcd/resource_vcd_catalog.go]: Yes, you are right. Fixed.

File: vcd/resource_vcd_catalog_test.go [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+		"CatalogName": TestAccVcdCatalog,
+		"Description": TestAccVcdCatalogDescription,
+	}
+
+	if vcdShortTest {
+		t.Skip(acceptanceTestsSkipped)
+		return
+	}
+	configText := templateFill(testAccCheckVcdCatalogBasic, params)
+	if os.Getenv("GOVCD_DEBUG") != "" {
```
[the remark this answers is not in the document]
Comment (vbauzysvmware) [reply vcd/resource_vcd_catalog_test.go]: Changed.

File: vcd/resource_vcd_catalog.go [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+				Required: true,
+				ForceNew: true,
+			},
+
+			"description": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"force": &schema.Schema{
```
[the remark this answers is not in the document]
Comment (vbauzysvmware) [reply vcd/resource_vcd_catalog.go]: Yes, changed that and added descriptions.

File: vcd/resource_vcd_catalog.go (hunk 3 above)
Comment (vbauzysvmware) [reply vcd/resource_vcd_catalog.go]: These options manages how delete should behave (delete subfolders/objects and should check state). If options are false and there are mentioned items then them hampers deletion and error returned.

File: vcd/resource_vcd_catalog.go [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+				Type:        schema.TypeBool,
+				Required:    true,
+				ForceNew:    false,
+				Description: "force=True along with recursive=True to remove a catalog and any objects it contains, regardless of their state.",
+			},
+			"delete_recursive": &schema.Schema{
+				Type:        schema.TypeBool,
+				Required:    true,
+				ForceNew:    false,
+				Description: "recursive=True to remove a Catalog and any objects it contains that are in a state that normally allows removal.",
```
Comment (lvirbalas) [inline vcd/resource_vcd_catalog.go]: Please adjust description to mention that this is for `destroy` operation as above.

BTW, which state do not "normally allow removal"? And will anything be deleted if there are such items in a catalog?

File: vcd/resource_vcd_catalog.go [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+				ForceNew: true,
+			},
+
+			"description": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: false,
+				Optional: true,
+				ForceNew: true,
+			},
+			"delete_force": &schema.Schema{
```
[the remark this answers is not in the document]
Comment (vbauzysvmware) [reply vcd/resource_vcd_catalog.go]: recursive and force is already in our org resource :)

File: vcd/resource_vcd_catalog.go (hunk 6 above)
Comment (vbauzysvmware) [reply vcd/resource_vcd_catalog.go]: Regarding BTW: normally any state do not allow removal. Recursive means delete catalog even items exists. Force allows to delete if item in not success state. Not exact definition in api. I tested with just failed upload.

Comment (vbauzysvmware) [reply vcd/resource_vcd_catalog.go]: Description updated.

File: vcd/resource_vcd_catalog.go [hunk 8]
```diff
@@ -43,13 +37,13 @@ func resourceVcdCatalog() *schema.Resource {
 				Type:        schema.TypeBool,
 				Required:    true,
 				ForceNew:    false,
-				Description: "force=True along with recursive=True to remove a catalog and any objects it contains, regardless of their state.",
+				Description: "when deleting force=True along with recursive=True to remove a catalog and any objects it contains, regardless of their state.",
```
Comment (lvirbalas) [inline vcd/resource_vcd_catalog.go]: Please adjust wording, capitalisation and use full parameter names:

> When destroying use delete_force=True with delete_recursive=True to remove a catalog and any objects it contains, regardless of their state.

File: vcd/resource_vcd_catalog.go [hunk 9]
```diff
@@ -43,13 +37,13 @@ func resourceVcdCatalog() *schema.Resource {
...
 			},
 			"delete_recursive": &schema.Schema{
 				Type:        schema.TypeBool,
 				Required:    true,
 				ForceNew:    false,
-				Description: "recursive=True to remove a Catalog and any objects it contains that are in a state that normally allows removal.",
+				Description: "when deleting recursive=True to remove a Catalog and any objects it contains that are in a state that normally allows removal.",
```
Comment (lvirbalas) [inline vcd/resource_vcd_catalog.go]: Please adjust wording, capitalisation and use full parameter names:

> When destroying use delete_recursive=True to remove the catalog and any objects it contains that are in a state that normally allows removal.

Comment (vbauzysvmware) [reply vcd/resource_vcd_catalog.go]: done

File: vcd/resource_vcd_catalog.go (hunk 8 above)
Comment (vbauzysvmware) [reply vcd/resource_vcd_catalog.go]: done

**Pull request merged**